### PR TITLE
[VPlan] Update VPInst::onlyFirstLaneUsed to check users.

### PR DESCRIPTION
**Pull request by fhahn**
A VPInstruction only has its first lane used if all users use its first lane only. Use vputils::onlyFirstLaneUsed to continue checking the recipe's users to handle more cases.

Besides allowing additional introduction of scalar steps when interleaving in some cases, this also enables using an Add VPInstruction to model the increment - as a follow up.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 1]
```diff
@@ -515,6 +515,24 @@ void VPInstruction::execute(VPTransformState &State) {
...
+  switch (getOpcode()) {
+  default:
+    return false;
+  case Instruction::ICmp:
+    return vputils::onlyFirstLaneUsed(this);
+  case VPInstruction::ActiveLaneMask:
+  case VPInstruction::CalculateTripCountMinusVF:
+  case VPInstruction::CanonicalIVIncrementForPart:
+  case VPInstruction::BranchOnCount:
+    return getOperand(0) == Op;
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:532]: This maintains current functionality, but leaves room for further improvement - to cover additional operands (of ActiveLaneMask, BranchOnCount) and opcodes (Not, Select, BranchOnCond) - worth a TODO.

File: llvm/test/Transforms/LoopVectorize/pr45679-fold-tail-by-masking.ll [hunk 2]
```diff
@@ -139,46 +139,42 @@ define void @pr45679(ptr %A) optsize {
...
+; VF1UF4-NEXT:    [[TMP4:%.*]] = icmp ule i32 [[TMP0]], 13
+; VF1UF4-NEXT:    [[TMP5:%.*]] = icmp ule i32 [[TMP1]], 13
+; VF1UF4-NEXT:    [[TMP6:%.*]] = icmp ule i32 [[TMP2]], 13
+; VF1UF4-NEXT:    [[TMP7:%.*]] = icmp ule i32 [[TMP3]], 13
+; VF1UF4-NEXT:    br i1 [[TMP4]], label [[PRED_STORE_IF:%.*]], label [[PRED_STORE_CONTINUE:%.*]]
 ; VF1UF4:       pred.store.if:
-; VF1UF4-NEXT:    [[INDUCTION:%.*]] = add i32 [[INDEX]], 0
-; VF1UF4-NEXT:    [[TMP4:%.*]] = getelementptr inbounds i32, ptr [[A:%.*]], i32 [[INDUCTION]]
-; VF1UF4-NEXT:    store i32 13, ptr [[TMP4]], align 1
+; VF1UF4-NEXT:    [[TMP8:%.*]] = getelementptr inbounds i32, ptr [[A:%.*]], i32 [[TMP0]]
```
Comment (ayalz) [inline llvm/test/Transforms/LoopVectorize/pr45679-fold-tail-by-masking.ll:153]: Noting: this is cse'ing TMP0 (as with TMP1,2,3 below), compared to currently re-computing it. In this case, should be dropped being a zero-add.

**Review by ayalz — APPROVED** [review]
LGTM, adding various comments.

The only observable changes to existing tests involve cse'ing IV steps from replicate regions when unrolling w/o vectorizing, iinm. Is there some other scenario worth testing?

"... this also enables using an Add VPInstruction to model the increment" - as a follow-up?

File: llvm/lib/Transforms/Vectorize/VPlan.cpp [hunk 3]
```diff
@@ -1397,9 +1397,9 @@ void VPSlotTracker::assignSlots(const VPBasicBlock *VPBB) {
...
 }
 
-bool vputils::onlyFirstLaneUsed(VPValue *Def) {
+bool vputils::onlyFirstLaneUsed(const VPValue *Def) {
   return all_of(Def->users(),
-                [Def](VPUser *U) { return U->onlyFirstLaneUsed(Def); });
+                [Def](const VPUser *U) { return U->onlyFirstLaneUsed(Def); });
 }
 
 bool vputils::onlyFirstPartUsed(VPValue *Def) {
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.cpp]: Done in 3444240540b0b36902dc8b9b11339b8969f7bca9, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 4]
```diff
@@ -1256,23 +1256,7 @@ class VPInstruction : public VPRecipeWithIRFlags {
     }
   }
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h:1258]: Added back, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:534]: Done, thanks!

**Comment by fhahn** [comment]
> The only observable changes to existing tests involve cse'ing IV steps from replicate regions when unrolling w/o vectorizing, iinm. Is there some other scenario worth testing?

I wasn't able to come up with a different test with vectorizing.

**Pull request merged**